### PR TITLE
Out of bound access on `test_sparse_sum_op`

### PR DESCRIPTION
**Pull request by lawrence910426**
### PR Category

Operator Mechanism

### PR Types

Bug fixes

### Description

1. `sum_kernel.cu`
The size of `out_cols_data` is only `x_dim0 * x_dim1`. It is illegal to access memory after x_dim0 * x_dim1. To prevent such illegal access, the loop in `SumCsr3DGradCudaKernel` is splitted into two loops. 

2. `sum_grad_kernel.cu`
The length of `x_crows_data` is only `x_dim0 * (x_dim1 + 1)`. Access to `x_crows_data[x_dim0 * (x_dim1 + 1)]` is in fact illegal. However, `x_crows_data[x_dim0 * (x_dim1 + 1)]` would be 0 to the alignment mechanism of `StreamSafeAllocator`.

Moreover, `dx_values_data` would never be filled when `index = x_dim0 * (x_dim1 + 1) - 1`. Therefore, the last iteration of the loop could be ignored.

### CODE REVIEW

File: paddle/phi/kernels/sparse/gpu/sum_grad_kernel.cu [hunk 1]
```diff
@@ -54,7 +54,7 @@ __global__ void SumCsr3DGradCudaKernel(const int64_t* x_crows_data,
                                        const int64_t x_dim1,
                                        T* dx_values_data) {
   // dout_crows_data[index] should be equal to number;
-  CUDA_KERNEL_LOOP_TYPE(index, x_dim0 * (x_dim1 + 1), int64_t) {
+  CUDA_KERNEL_LOOP_TYPE(index, x_dim0 * (x_dim1 + 1) - 1, int64_t) {
```
Comment (zhwesky2010) [inline paddle/phi/kernels/sparse/gpu/sum_grad_kernel.cu:57]: 这个是因为对于3d csr，crows的长度为x_dim0 * (x_dim1 + 1)

CUDA_KERNEL_LOOP_TYPE中只会访问到num-1，这样写应该不会触发到边界问题

infoflow 2024-05-10 12-19-55

File: paddle/phi/kernels/sparse/gpu/sum_kernel.cu [hunk 2]
```diff
@@ -137,11 +137,16 @@ __global__ void SumCsr3DCudaKernel(const int64_t* x_crows_data,
                                    int64_t* out_crows_data,
                                    int64_t* out_cols_data,
                                    T* out_values_data) {
+  {
+    CUDA_KERNEL_LOOP_TYPE(index, x_dim0 * x_dim1, int64_t) {
+      out_cols_data[index] = 0;
```
Comment (lawrence910426) [inline paddle/phi/kernels/sparse/gpu/sum_kernel.cu:142]: The size of out_cols_data is only x_dim0 * x_dim1. It is illegal to access memory after x_dim0 * x_dim1.

File: paddle/phi/kernels/sparse/gpu/sum_grad_kernel.cu (hunk 1 above)
Comment (lawrence910426) [reply paddle/phi/kernels/sparse/gpu/sum_grad_kernel.cu:57]: 在 Line 68 的 `x_crows_data[index + 1]` 会触发 out of bound，所以必须 -1

**Review by zyfncg — APPROVED** [review]

**Comment by luotao1** [comment]
hi, @lawrence910426 
- 非常感谢你对飞桨的贡献，我们正在运营一个PFCC组织，会通过定期分享技术知识与发布开发者主导任务的形式持续为飞桨做贡献，详情可见 https://github.com/luotao1 主页说明。
- 如果你对PFCC有兴趣，请发送邮件至 [ext_paddle_oss@baidu.com](mailto:ext_paddle_oss@baidu.com)，我们会邀请你加入~

**Pull request merged**